### PR TITLE
Add scroll behaviour for container elements

### PR DESCRIPTION
**Pull request by Mesoptier**
[_Dual of this PR on react-router-scroll_](https://github.com/taion/react-router-scroll/pull/27)

This will allow you to register scroll container elements with scroll-behavior. I abstracted most of the behavioural logic from the `ScrollBehavior` class to a `Container` class (I am not certain about naming), leaving the `ScrollBehavior` class to be a manager for the containers.

API changes:

- Added `ScrollBehavior#registerContainer(key, domNode)`
- Added `ScrollBehavior#unregisterContainer(key)`
- Added `ScrollBehavior#getContainerKeys()` to return an array of container keys.
- `ScrollBehavior#stop()` now stops and unregisters all containers.
- `ScrollBehavior#updateScroll(key, scrollPosition)` now requires a container key.
- `shouldUpdateScroll(prevLocation, location, key)` now receives an additional container key parameter. The function will be called for each container so you have more control over which containers will be updated.

I have not yet found a way to neatly expose the register/unregister API with this package. But it already works nicely with `react-router-scroll`.

### CODE REVIEW

**Comment by taion** [comment]
Ohhh this is awesome and it's exactly the approach I wanted to suggest.

In terms of the API here, I'd just add `registerScroll` and `unregisterScroll` methods to the history. Do you think that would work?

File: src/ScrollBehavior.js [hunk 1]
```diff
@@ -15,21 +16,22 @@ const KEY_PREFIX = 's/';
 // Try at most this many times to scroll, to avoid getting stuck.
 const MAX_SCROLL_ATTEMPTS = 2;
 
-export default class ScrollBehavior {
-  constructor(history, getCurrentLocation) {
+function scrollTo(node, x, y) {
+  if (isWindow(node)) {
+    node.scrollTo(x, y);
+  } else {
+    node.scrollLeft = x;
```
Comment (taion) [inline src/ScrollBehavior.js]: ESLint pragma this.

**Comment by Mesoptier** [comment]
Yes that would work, however I think a cleaner way would be to instead add `scrollBehavior` to the history.

**Comment by taion** [comment]
Hmm. I think most of the methods on `ScrollBehavior` are not intended to be callable by the user, though – the wrapper for `history` or React Router should take care of e.g. `start` and `stop`, and in this context I think it'd be better if those methods were not directly exposed to the user.

**Comment by Mesoptier** [comment]
My thinking behind only exposing `scrollBehavior` is to not pollute the history object. I think that it should be absolutely clear that this is an extension on the history object. We could take the best from both worlds and add something like this: 

```javascript
scrollBehavior: {
  registerScroll: scrollBehavior.registerScroll,
  unregisterScroll: scrollBehavior.unregisterScroll,
},
```

That way the functions are "namespaced", and still only the relevant functions are exposed.

**Comment by taion** [comment]
I think it's just one property versus two, though.  `registerScroll` and `unregisterScroll` are unlikely to collide with anything else. `registerScroll` can even return an `unregister` callback, maybe.

In all honesty, I don't really expect much of anybody to use this library directly – those hooks are just so we can write tests here, mostly.

**Comment by Mesoptier** [comment]
Fair enough, I will add both functions and the relevant tests tomorrow!

**Comment by taion** [comment]
What do you think of not having an explicit `history.unregisterScroll`, and just having `history.registerScroll` return a hook to unregister? Seems like that reduces the API surface area here.

**Comment by Mesoptier** [comment]
If we're going to add `registerScroll`, I think we should also add `unregisterScroll` because 1) that matches the API on the scrollBehavior object and 2) it is much easier to keep track of a key than it is of a function.

BTW, I was thinking about naming the methods `(un)registerScrollContainer` instead

**Comment by taion** [comment]
@Mesoptier 

I think that's a good point – the issue though is that the rest of the history API is `listen` and `listenBefore` returning "unlisten" callbacks, rather than having explicit `unlisten` and `unlistenBefore` methods.

It's not that I think "registerScroll(Container)" returning a callback is better objectively – it's just more consistent with `listen` and `listenBefore`.

**Comment by taion** [comment]
@Mesoptier Would you mind if I merged with this and ran with the API nits on my own? Thanks for this code!

**Comment by Mesoptier** [comment]
Yeah, of course, go ahead. And you're welcome! :smile: 

BTW, some tests did not pass on my local Firefox installation but did pass on Travis. That might be something to look into. (I have the latest Firefox for Windows)

**Pull request merged**